src/porepy/numerics/ad/secondary_operator.py

```diff
@@ -0,0 +1,897 @@
+           returned.
+
+        Parameters:
+            domains: A subset of either grids, mortar grids or boundary grids in the
+                md-grid passed at instantiation.
+
+        Raises:
+            ValueError: If ``domains`` is not composed of either grids, mortar grids or
+                boundary grids.
+
+        """
+
+        # This is for completeness reasons, when calling equations on empty list
+        if len(domains) == 0:
+            return pp.wrap_as_dense_ad_array(np.zeros((0,)), name=self.name)
+        # On the boundary, this is a Time-Dependent dense array
+        elif all(isinstance(g, pp.BoundaryGrid) for g in domains):
+            return pp.ad.TimeDependentDenseArray(self.name, domains)
+        # On subdomains or interfaces, create the secondary operators
+        elif all(isinstance(g, pp.Grid) for g in domains) or all(
+            isinstance(g, pp.MortarGrid) for g in domains
+        ):
+            # for mypy
+            domains_ = cast(Sequence[pp.Grid] | Sequence[pp.MortarGrid], domains)
+            children = [child(domains_) for child in self._dependencies]
+
+            # Check if first-order dependency
+            assert all(isinstance(child, pp.ad.Variable) for child in children), (
+                "Secondary expressions must depend on independent variables, not"
+                + f" {[type(c) for c in children]}."
+            )
+
+            # always start with operator at current time step, current iterate
+            op = SecondaryOperator(
+                name=self.name,
+                domains=domains_,
+                children=children,
+            )
+
+            # assign the function which extracts the data
+            op.fetch_data = self.fetch_data
+
+            return op
+        else:
+            raise ValueError(
+                f"Unsupported domain configuration {[type(g) for g in domains]}."
+            )
+
+    def fetch_data(
+        self, op: SecondaryOperator, grid: pp.GridLike, get_derivatives: bool = False
+    ) -> np.ndarray:
+        """Function fetching the data stored for this secondary expression, represented
+        in Ad form by ``op``.
+
+        ``op`` has a time step or iterate index, which specifies which data should be
+        fetched.
+
+        This function is assigned to :attr:`SecondaryOperator.fetch_data`
+        (a work-around such that the Ad operator has no reference to the md-grid like
+        the other operators).
+
+        Note:
+            This method uses PorePy's utility functions to get data stored in a grid's
+            data dictionary. Respective errors will be raised if no data stored.
+
+        Parameters:
+            op: An AD representation of this expression at a specific time and iterate.
+            grid: A grid in the md-grid for which the data should be fetched.
+            get_derivatives: If True, it fetches the derivative values, otherwise it
```

Comment:
Should we explicitly raise an error if the derivative is requested for an operator which is not at the current time step? I see from the test that this will result in a failed lookup in the data dictionary, hence a `KeyError` will be raised, but that will not be very helpful for a user.

Reply:
I left this, just in case that for some the user stores the derivative values at previous time or iterates.
Maybe an error would be better, because Variables evaluated at prev time or iter give an AD array with jac=0.
It would theoretically trigger the `get_jacobian` in `SecondaryOperator`, and there the `KeyError` is raised.
But conceptually, even without `KeyError`, we cannot use the previous derivatives because the variables have `jac=0`.

Will that ever change?

Reply:
I don't think we can rule out that, in the future, evaluation at a previous time step or iterate may return a numpy arrays instead of an Ad array.

Reply:
Added `ValueError`, if `op` is at previous time or derivative and the user attempts to get derivative values for it.